frappe/utils/oauth.py

```diff
@@ -237,110 +225,96 @@ def login_oauth_user(
-		save = True
-		user.set_social_login_userid(provider, userid=data["sub"])
-
-	elif provider == "salesforce" and not user.get_social_login_userid(provider):
-		save = True
-		user.set_social_login_userid(provider, userid="/".join(data["sub"].split("/")[-2:]))
-
-	elif not user.get_social_login_userid(provider):
-		save = True
-		user_id_property = frappe.db.get_value("Social Login Key", provider, "user_id_property") or "sub"
-		user.set_social_login_userid(provider, userid=data[user_id_property])
-
-	if save:
+
+def update_oauth_user(user: str, data: dict, provider: str):
+	if isinstance(data.get("location"), dict):
+		data["location"] = data["location"].get("name")
+
+	user: "User" = get_user_record(user, data)
+	update_user_record = user.is_new()
+
+	if not user.enabled:
+		frappe.respond_as_web_page(_("Not Allowed"), _("User {0} is disabled").format(user.email))
+		return False
+
+	if not user.get_social_login_userid(provider):
+		update_user_record = True
+		match provider:
+			case "facebook":
+				user.set_social_login_userid(provider, userid=data["id"], username=data.get("username"))
+				user.update({"user_image": f"https://graph.facebook.com/{data['id']}/picture"})
+			case "google":
+				user.set_social_login_userid(provider, userid=data["id"])
+			case "github":
+				user.set_social_login_userid(provider, userid=data["id"], username=data.get("login"))
+			case "frappe" | "office_365":
+				user.set_social_login_userid(provider, userid=data["sub"])
+			case "salesforce":
+				user.set_social_login_userid(provider, userid="/".join(data["sub"].split("/")[-2:]))
+			case _:
+				user_id_property = (
+					frappe.db.get_value("Social Login Key", provider, "user_id_property") or "sub"
+				)
+				user.set_social_login_userid(provider, userid=data[user_id_property])
+
+	if update_user_record:
 		user.flags.ignore_permissions = True
 		user.flags.no_welcome_mail = True
 
-		# set default signup role as per Portal Settings
-		default_role = frappe.db.get_single_value("Portal Settings", "default_role")
-		if default_role:
+		if default_role := frappe.db.get_single_value("Portal Settings", "default_role"):
 			user.add_roles(default_role)
 
 		user.save()
 
 
-def get_first_name(data):
+def get_first_name(data: dict) -> str:
 	return data.get("first_name") or data.get("given_name") or data.get("name")
 
 
-def get_last_name(data):
+def get_last_name(data: dict) -> str:
 	return data.get("last_name") or data.get("family_name")
 
 
-def get_email(data):
+def get_email(data: dict) -> str:
 	return data.get("email") or data.get("upn") or data.get("unique_name")
 
 
-def redirect_post_login(desk_user, redirect_to=None, provider=None):
-	# redirect!
+def redirect_post_login(desk_user: bool, redirect_to: str | None = None, provider: str = None):
 	frappe.local.response["type"] = "redirect"
 
 	if not redirect_to:
-		# the #desktop is added to prevent a facebook redirect bug
 		desk_uri = "/app/workspace" if provider == "facebook" else "/app"
```

Comment:
this might error out if provider is not present?

`None` compared with `str` (?)

Reply:
![batman-thinking](https://user-images.githubusercontent.com/36654812/210200088-04b52d0a-7d53-4022-b261-52e26bd194a9.gif)

not sure I follow what your concerns really are here

Reply:
I guess this whole function can be written as:

```py
def redirect_post_login(desk_user, redirect_to=None):
	if not redirect_to:
		redirect_to = frappe.utils.get_url("/app" if desk_user else "/me")

	frappe.local.response["type"] = "redirect"
	frappe.local.response["location"] = redirect_to
```

`provider` param's default value doesn't make sense and in the end it's anyway redirecting to `/app` :P

Reply:
> provider param's default value doesn't make sense (should be "") 

I don't understand this line of reasoning tbh. Can you elaborate? Empty string and None are not the same - lack of provider should be None not "". Are you trying to promote a stringily typed system? 😢 

> and not sure why it's redirecting to workspace list for facebook

I think this was a residual product of the routing change [/ and after some searching a series of ambiguous commits]. Not sure this PR is the best time to see if this condition should be changed 🤣

Reply:
my bad - thought `None` compared with `str` would've errored out